### DISCUSSION TITLE
Is there a pure DuckDB way to order GeoParquet output spatially?

### DISCUSSION THREAD

**Discussion by cholmes** [duckdb/duckdb-spatial#419] [Q&A] [answered]
So with DuckDB 1.1 there's GeoParquet output, with bbox struct (awesome), and r-tree indexes (also awesome). But I loaded some geospatial data that wasn't ordered and then wrote out the geoparquet and it's clearly not indexed / ordered at all. Is there a way to tell DuckDB to order the data by the r-tree?  I tried creating an r-tree index and then doing 'order by geometry', but it didn't seem to do anything.

My default way would be to create a quadkey and order by that - but it seems like the native r-tree should be able to help me and hopefully remove a couple steps.

Happy to contribute to the docs / share broadly if there's a way to do it.

**Comment by Maxxen** [comment, marked as answer]
Hello! Unfortunately, DuckDB does not support overriding the comparison operators (used to sort) for custom types, meaning `ORDER BY <geometry>` will just order based on the lexical order of the bytes in the internal binary representation. The r-tree index is completely separate from ordering. I know that other databases (like Postgres) sometimes can perform an index scan instead of explicitly sorting when executing an order-by, but this requires the ordering criteria to map to the physical order of the index. In the case of R-Trees, I can't think of any way that would work. 

Perhaps there could be a "dummy" scalar function the optimizer could detect to do this if the column is r-tree indexed, or otherwise error out, e.g. `SELECT geom FROM tbl ORDER BY rtree_order(geom)`. But again, the r-tree traversal order might not actually be the most efficient to use for spatial clustering, compared to e.g. a Hilbert curve. Additionally, doing index lookups for every row in the table is going to be significantly more expensive than a full table scan. DuckDB indexes don't store any actual data within themselves (unlike how indexes sometimes work in other databases), but just the `row_id` of the row in the indexed table. This means that every time an index scan is performed, DuckDB first has to traverse the index, collect the row ids, and then perform (what is most likely) random-access point-lookups from the underlying table row-by-row. 

There are also some other issues related to the preservation (or lack thereof) of row-order when scanning from a table in parallel or writing to parquet. I think the only way to guarantee the output order of the rows within parquet files is to execute a ORDER BY, which will emit chunks with a "batch index", that the parquet writer can use to reorganize incoming chunks in flight and ensure they are written in sorted order, even when executing with multiple threads. 

So uh, yeah, we could look into adding some sort of dummy marker function + optimizer rule + index-batch-scan operator, but I don't think its going to be much faster or produce much better results than just doing a full table scan + explicit sort, e.g. `COPY (SELECT geom FROM tbl ORDER BY ST_Hilbert(geom) ) TO 'gpq.parquet`

**Reply by cholmes** [reply]
Cool. That's all great to know. I think all I was really looking for was your last line - I didn't realize there was ST_Hilbert.

**Reply by Maxxen** [reply]
By default it maps the entire space of two floating point coordinates into a single uint32, which produces kind of messy Hilbert curves, but it is overloaded so you can pass an optional "bounds" parameter which will normalize the centroid of the geometry to 0-1 based on how close it maps to the boundary, which gives the resulting Hilbert curve better precision.

**Reply by cholmes** [reply]
Ok, finally coming back to this - I have a real use case for this now. I was wondering if my results are as they should be:
![duckdb-hilbert2](https://github.com/user-attachments/assets/5bf612d0-91a3-4fe5-a483-3331d13518e8)

It certainly looks cool. I'm used to seeing it more as like big chunks loading and then moving to the next, so just wanted to be sure that this output is what's expected, and that it performs well for spatial queries. I just did `ORDER BY ST_Hilbert(geometry)`. I wasn't sure how the bounds parameter helps - would it be better if I did `ORDER BY ST_Hilbert(geometry, ST_Envelope(geometry))`? Or are you saying use some other bounds information to make it better?

**Reply by Maxxen** [reply]
Im not great at converting the exact math terminology, but basically, without a bounds you're effectively trying to "map" values (coordinates) within an interval of `(f32min....f32max)^2` into an interval of  `(u32min...u32max)` which is going to be very lossy. But if you pass a `bounds` parameter all coordinates are first normalized to within the bounds and then Hilbert-encoded, which will most likely give you a much tighter packing as the source interval is reduced to `((bounds_xmin...boundsxmax) * (bounds_ymin...bounds_ymax))`. Therefore the bounds parameter should encompass the whole dataset, otherwise you're going to map geometries to different Hilbert curves and their Hilbert values will most likely not be spatial related.

**Reply by Maxxen** [reply]
Here's two examples. Unfortunately I don't have a great GIS/viewer at hand so the colors are quantized instead of a proper hue ramp, but It should get the idea across. With a tighter bound you get less noise and better spatial locality.

## Without bounds
![image](https://github.com/user-attachments/assets/17a51179-b90c-4732-8c0a-035876bd814c)


## With bounds (-180, -90, 180, 90)
![image](https://github.com/user-attachments/assets/c002b2fa-0b14-4d09-9454-7858e2a337dd)

**Reply by Maxxen** [reply]
(and the code)

No bounds
```
CREATE TABLE t1 as SELECT x::GEOMETRY as geom, st_hilbert(x::GEOMETRY)::BIGINT as val
FROM ST_GeneratePoints({min_x : -180, min_y : -90, max_x: 180, max_y : 90}::BOX_2D, 100000) as t(x);

COPY t1 TO 'test1.geojson' WITH (FORMAT GDAL, DRIVER GeoJSON);
```

With bounds
```
CREATE TABLE t1 as SELECT x::GEOMETRY as geom, st_hilbert(x::GEOMETRY, {min_x : -180, min_y : -90, max_x: 180, max_y : 90}::BOX_2D)::BIGINT as val
FROM ST_GeneratePoints({min_x : -180, min_y : -90, max_x: 180, max_y : 90}::BOX_2D, 100000) as t(x);

COPY t1 TO 'test2.geojson' WITH (FORMAT GDAL, DRIVER GeoJSON);
```

**Reply by cholmes** [reply]
Ah, I see! I was struggling to understand how to use the bounds - didn't realize it was supposed to be for the full dataset. Changed the call to `ORDER BY ST_Hilbert(geometry, ST_Extent(ST_MakeEnvelope(-180, -90, 180, 90)))` and got results much more like I was expecting:
![duckdb-hilbert-fixed](https://github.com/user-attachments/assets/f4581e34-34c9-4356-9c05-37b3d4224730)

Thanks! For this dataset it's global, so I could just use the total world's bounds. And then for other datasets it seems like you just need to do like `SELECT ST_Extent_Agg(geometry) AS dataset_bounds FROM <tablename>` to get the bounds and then you could pass that in? 

Is there a reason to not provide that as a short cut in DuckDB directly? Like it seems like the most common case for the bounds for the hilbert would just be the bounds of the dataset?

**Reply by Maxxen** [reply]
Yes! It's mostly a limitation of how the query engine is structured. You have to make a pass over every input geometry to calculate the bounds and _then_ another pass to calculate the Hilbert code now that the bounds are known, but that's not possible with a normal scalar function. And even if it was, because DuckDB is a streaming engine you'd have to effectively stop the query pipeline and buffer the whole queryset (potentially to/from disk) which would be pretty expensive. 

If extensions could provide custom statistics we could potentially keep the known bounds around for each column, but that's not possible yet. However, I have some ideas on how to manage/track SRID/projection info for each column id like to get to in the near future, and depending on how that goes we might be able to automatically infer the bounds as the bounds of the coordinate system.

**Reply by Maxxen** [reply]
Also, there isn't really a way in SQL to pass aggregates as input to scalar functions without separating it into multiple statements (first compute the aggregate result, e.g. st_extent_agg) and then pass it to the scalar (st_hilbert). So we can't create a macro or query-rewrite-level shortcut either.

**Reply by cholmes** [reply]
Ah, got it. That all makes a lot of sense. Could it be a mini-extension? Like is there a way with an extension to just have it run the two statements? Anyways, I'm happy to document it in the [geoparquet publishing best practices](https://github.com/opengeospatial/geoparquet/pull/254/files) I'm working on. And would be happy to contribute to the docs or a make a little tutorial. It's pretty simple, but it's not obvious at all from the info I could find.

**Reply by Maxxen** [reply]
The closes I can think of is a table macro, like this
```sql
CREATE OR REPLACE FUNCTION hilbert_auto(col_name, table_name) 
AS TABLE SELECT 
  st_hilbert(COLUMNS(col_name), (
    SELECT st_extent(ST_Extent_Agg(COLUMNS(col_name)))::BOX_2D FROM query_table(table_name))
  ) FROM query_table(table_name);
```
You'd use it like
```
SELECT * FROM hilbert_auto('geom', 't1');
```
This queries the table given by 'table_name' twice, once to compute the bounds and once to get the Hilbert values.

> It's pretty simple, but it's not obvious at all from the info I could find.

Yeah this would for sure be good blog-post material! I want to add more guide/tutorial style content to docs at some point as well but I want to get some more stuff ready code-side before digging into that.

**Reply by cholmes** [reply]
The hilbert_auto didn't quite work for me, as it seemed to just return a single column. I asked chatgpt and got:
```
SELECT *
  FROM fsq
  ORDER BY st_hilbert(
      geom,
      (
          SELECT st_extent(ST_Extent_Agg(COLUMNS(geom)))::BOX_2D
          FROM fsq
      )
  )
```

Which seemed to work? But only tested on ~5 million rows. Not sure if there's something wrong with that approach? It also suggested:
```
WITH bounds_cte AS (
  SELECT st_extent(ST_Extent_Agg(COLUMNS(geom)))::BOX_2D AS bounds
  FROM fsq
)
SELECT f.*
FROM fsq AS f
ORDER BY st_hilbert(
    f.geom,
    (SELECT bounds FROM bounds_cte)
);
```
which I didn't try out, and it didn't seem more readable to me.

> Yeah this would for sure be good blog-post material!

Cool, I'll try to throw something up. Let me know which of the above to do, or else another way to do it with two lines of SQL.

> I want to add more guide/tutorial style content to docs at some point as well but I want to get some more stuff ready code-side before digging into that.

Whenever you start this I'd be happy to try to add a few sections for some of the things I do often.
